### ISSUE TITLE
Liquid system improvements

### ISSUE THREAD

**Issue by MarkuBu** [minetest/minetest#5387]
The liquid system in Minetest is strange. A single source placed on the ceiling causes a cross shaped stream with 5 lines. I would expect 1

![screenshot_20170313_175334](https://cloud.githubusercontent.com/assets/16163058/23867551/870abb64-081c-11e7-975c-9e088d877e52.jpg)

And a single source is able to flood a half world. Yes, this is caused by the single lava source at the top.

![screenshot_20170313_183853](https://cloud.githubusercontent.com/assets/16163058/23867659/eaeaefbe-081c-11e7-82bd-91067fe516f5.jpg)

And real liquids don't flow "around the corner"

![screenshot_20170313_193312](https://cloud.githubusercontent.com/assets/16163058/23869731/4f71052a-0824-11e7-9531-537deb375555.jpg)

Real Liquid source have a limited output per time and flow downhills, but not around corners. Well, they can if there is no other way.

**Comment by tenplus1** [comment]
I would agree with this one, liquids should fall downward and only spread when they hit a node below.

**Comment by MarkuBu** [comment]
They can spread to all sides if there is no way down.

The problem is: liquid source is an unlimited source to all directions, but liquid flowing is also an unlimited source to all directions, just with a lower range

**Comment by MarkuBu** [comment]
And water does not spread on water. It just flows in

![screenshot_20170313_221008](https://cloud.githubusercontent.com/assets/16163058/23875471/e4ece51e-0839-11e7-97f0-d4916eb784c7.jpg)

**Comment by red-001** [comment]
Apparently water spreading on water is a feature not a bug as the river mapgen uses it to create slopes

**Comment by Wuzzy2** [comment]
While these are all valid complaints, this can't be fixed simply by changing how liquids work.
Any major changes can not be done lightly without risking to break existing worlds. Since liquids are very common in many worlds.

I think the first issue is really a matter of taste and debatable. It should be up to subgames to decide.

That liquid sources don't fall might seem dumb, but we must not disable it by default in order to keep existing worlds safe. It is a common use case to use those “floating” liquid sourced to build something which looks like fountains. This change would break fountains as they crumble together after and update.
Such a change would be nearly as bad as turning all solid nodes like stone into falling nodes by default.

That liquid sources can flood entire worlds is a valid concern. But again, it must not be changed for compability with existing worlds. If someone built a nice looking structure where the exact flow of the liquid is very important for artistic reasons, changing the liquid flowing range even slightly would destroy it. Think of a complex lava river.
But it is still a valid concern because it is undesired gameplay for some subgames. Maybe one possibility to improve this is to add an internal “liquid parameter” (maybe in `param2`?) to indicate the *total* flowing distance from the nearest liquid source feeding into it. Flowing liquids could then simply stop flowing if the *total* flowing distance is exceeded, even if they would otherwise continue to flow under the normal rules.
But this would be a MAJOR change from liquid mechanics, therefore such a limit must be optional and infinite by default for compability.
A much simpler solution which already works is by reducing the normal flowing range of the liquid. A flowing range of 8 is very high.

> And liquids don't flow "around the corner"

I have no idea what you're talking about here. In the screenshots, the water actually flows around the corner, directly contradicting your statement. I'm confused.

> Apparently water spreading on water is a feature not a bug as the river mapgen uses it to create slopes

I agree, I don't see a problem here either. Also, no alternative for a different liquid mechanic was suggested. This is not helping.

---

The possible problems I wrote above can be avoided if liquid definitions become more powerful, so that subgames can choose how their liquids should work

Specifically:
* Liquid sources fall down on/off
* Liquid sources flow downwards in …: 5 blocks / only 1 block
* Hard flowing harge limid for flowing liquids: 1 block / 2 blocks / … / infinite

Another benefit of this would be that you could have e.g. floating liquid sources and “falling” water sources in the same game. And it allows for greater variety.

Finally: If you don't like how liquids work, PLEASE be prepared to suggest an alternative gameplay mechanic at least. Then we have something to discuss. ;-)

**Comment by MarkuBu** [comment]
With "liquids don't flow around a corner" i mean real liquids. If water can't go anywhere else, then it can also flow around corners, but not in this situation.

The argument with breaking existing world's is futile. We already changed leaf decay, which breaks each world with non default trees and any mod that adds new trees.

And everybody can save a copy of an older version of Minetest. Or we can add some kind of legacy liquids

Just to rescue some rivers and fountains in some old mods is no reason for me to go forward.

Just my two cents to this topic

I have some ideas, but because my C++ skills are not the best i would like to make some experiments in lua, but I'm not sure if this is possible

**Comment by MarkuBu** [comment]
I forgot: I never complained about non falling liquid nodes

**Comment by MarkuBu** [comment]
I found an interesting video about Minecraft water. Maybe we can start to talk about this

https://www.youtube.com/watch?v=9uIRD9sMBU0

The interesting part begins at around 6:30

It is not far from Minetest water. The biggest difference is, that it try's to run down slopes instead of flooding everything

My biggest complain about Minetest water is, that it is flooding everything. It is not possible to make natural looking waterfalls or similar things without building a dam.

With a more natural behavior (water flows down a hill) it would also stop flooding caves

**Comment by paramat** [comment]
Please could you clarify your first post by referring either 'real' liquid or game liquid in each issue?

Much of this has been discussed before. Water spreading out on water is essential for river slopes.

It may be possible for a singe source to flow down instead of spreading out first, it may also be dificult.

Real liquids do flow around corners, but they have momentum which means they have a slight preference for maintaing their course.

Sources not being infinite, we've had 'finite water' before now and it was very intensive, you would need to define infinite sources for the source of a river, because everything is flowing there is no steady state so the processing load is huge.

Flowing towards the most downhill direction instead of spreading out may be possible, i agree the extreme spreading is a problem.

Overall, it is unrealistic of course, this is no issue, it has to be to be lightweight and simple, all of MT is like this, MC too. they are not physics simulations and are not realistic. Designing a liquid flow system is very difficult and often behaviour is how it is for good reason.

**Comment by MarkuBu** [comment]
Edit first post. Hope that clarify everything.

I think, Minetest water is not far away from MC water. The biggest difference (and biggest issue) of MT water is, that it spreads to all directions

I have an idea I would like to try. If a water source is generated a search for air is made at y - 1 in a diamond shape. If air is found the pathfinder could be called for the route and water is placed along the route

I don't know how MT water exactly works. MC water has two states, static and flowing (both water source and water flowing can have this states) and only flowing state is been updated at each tick. Static state is ignored which saves a lot of time.

It's similar to gravel and sand. They float in the air until a node update

**Comment by paramat** [comment]
When MT liquid falls down in a column and touches ground this is actually a flowing node, so flowing nodes also need to search around themselves for the most downhill direction.

**Comment by MarkuBu** [comment]
Yes, but even these flowing nodes become static once they are generated. Just like sand. You don't check each second if the sand can fall or not. Only if a neighbor node triggers a node update the sand node checks if it can fall again

Please don't mix static/flowing **state** with water source and water flowing **nodes**

**Comment by MarkuBu** [comment]
I'm sorry, it's called "still" water, not static. This video describes it very well

https://www.youtube.com/watch?v=MyeW-8IsZ-A

**Comment by sorcerykid** [comment]
Just to chime in, but if any substantial changes are made to the liguid generation algorithm, I do hope there will be a means to enable legacy support to ensure the integrity of existing builds.

**Comment by C1ffisme** [comment]
I suppose it's too late for this huge of a change, but here's my idea for liquid physics:

1. Liquid is one node with 8 "levels".
2. When you place a liquid (Or when liquids are 'ticked'), it will first check underneath for air or breakable nodes, and fall down if there is. Otherwise, it will stay where it is.
3. Next, it will check to the sides. If there is no air or breakable nodes in the four cardinal directions, it will remain static. If there is, it will give away levels to each area. EG:

Tick 1: (Air)(Water:8)(Air)

Tick 2: (Water:1)(Water:6)(Water:1)

4. If it detects multiple blocks of water of a lower level, and the water node has enough to evenly divide between all of them, it will give each one a level of water. If it cannot be divided evenly, it will try and give to the one with the lowest level if the difference between the two is greater than 1. Example:

Tick 1: (Air)(Water:8)(Water:5)

Tick 2: (Water:1)(Water:6)(Water:6)

Tick 3: (Water:2)(Water:5)(Water:6)

Tick 4: (Water:3)(Water:5)(Water:5)

Tick5: (Water:4)(Water:4)(Water:5)

5. Liquid with 1 level will not spread at all.
6. Buckets can only pick up a liquid with at least 8 levels.
7. If a liquid detects another liquid below it with a lower water level, it will take 1 from it's own level it's own level and fill the one below:

Tick 1: 
* (Water:6)(Block)
* (Water:8)(Air)

Tick 2:
* (Water:5)(Block)
* (Water:8)(Water:1)

Did I miss anything?

**Comment by MarkuBu** [comment]
No sure if I understand what you mean, but your approach sounds more difficult to add to MT than mine

**Comment by paramat** [comment]
> Yes, but even these flowing nodes become static once they are generated

Yes i understand that and did not say they were not static. I was just mentioning that flowing nodes need to search around them too in this idea:

> If a water source is generated a search for air is made at y - 1 in a diamond shape. If air is found the pathfinder could be called for the route and water is placed along the route

**Comment by MarkuBu** [comment]
> I was just mentioning that flowing nodes need to search around them too in this idea:

I'm not 100% sure about this. As I understand this, only the initial water source makes a search

**Comment by MarkuBu** [comment]
Probably falling water searches for slopes if it hit's a solid node.

**Comment by paramat** [comment]
Yes, to avoid huge spreads flowing water above a solid node will need to search for slopes, because there can be a single source node at the top of a huge flow and everyting below is water_flowing.

**Comment by MarkuBu** [comment]
I want to make a proof of concept, but a little bit simplified. I define nodebox nodes for different water levels instead of liquid nodes. I just want to simulate the flowing of minecraft water

**Comment by MarkuBu** [comment]
Damn, I can't use find_path because pathfinder doesn't search over air

**Comment by C1ffisme** [comment]
@MarkuBu 
> No sure if I understand what you mean, but your approach sounds more difficult to add to MT than mine

Yeah, it's closer to Terraria's liquid system than Minecraft's. It makes finite liquids a bit more realistic. I'm not sure how you'd deal with infinite liquids, though. (Probably not at all.)

**Comment by Avamander** [comment]
I think the cross shaped stream down is the thing that seems the thing that could be fixed the fastest, in addition probably also decrease the extreme flooding and weirdness of the liquids. Indeed this again can be turned into a flag, but to be honest I'd make 1 block stream the default, this ain't no Linux kernel where preserving user's userspace is really important, this is a game, things change, weird things should even more so.

**Comment by Wuzzy2** [comment]
Seeing arguments both pro and contra falling liquids I think it is safe to say that it would be best to let subgames decide. Why not support both? Both have legitimate use cases.
Allow to change it per-node, so you could have both falling and static liquids in the same game.

The same applies from the strange 5 cross water split.

**Comment by C1ffisme** [comment]
@Avamander Yep. See point 2 of my system. I other thing Terraria's liquids do, though, is they seem to be able to create waterfalls that don't place or remove water. I'm not quite sure how that works.

**Comment by HybridDog** [comment]
> The issue with fixing water spreading on water is that the valleys mapgen relies on it to create slopes. It is possible to generate these slopes without this bug but it is a bit more complicated then just making water flow straight down.

You can simply disable flowing water spreading on water source if the node which causes the spreading water has a flowing node above it.
The slopes aren't slopes in this case.

#
Differentiating between active and still water could fix the water flooding without breaking compatibility:
When adding a water source:
 Water source is placed, set to active
When adding a flowing water:
 Flowing water is placed, set to active
When a node is set to active and a time delay is over:
 The node is set to still, 
 if there's place for a new flowing water node, it is added
 else if the node is flowing water, the previous flowing water or source is set to active

BTW: The single water source flooding everything has another problem: If you remove the water source node, you have to load all the blocks which contain flowing water, else the flowing water on the lower end continues spreading everywhere although the source doesn't exist.

**Comment by MarkuBu** [comment]
Want to post my idea here again because it is off topic in #6755

Instead of spread to any direction it would make sense that the nodes around are checked if the liquid can flow down. If that is possible on one or more sides the liquid spreads only downwards, if not it spreads like now to all directions.

I would try to implement this myself, but I don't even know in which file I should start

**Comment by paramat** [comment]
It's all here, have fun https://github.com/minetest/minetest/blob/20a85d76d94c9c5c7fbe198c3d0e1fbee97a485f/src/map.cpp#L545

**Comment by MarkuBu** [comment]
> It's all here, have fun 

If I would know more about the Minetest core and with better C++ skills... probably.

**Comment by MarkuBu** [comment]
Nope, sorry. I tried it and I have a rough idea how it could work, but I don't know enough C++ and I'm not familiar enough with MT core to implement it

**Comment by paramat** [comment]
It's difficult code, not sure i could code it either.

Concerning flowing downhill https://github.com/minetest/minetest/issues/6755#issuecomment-350576820 the suggestion (which is a good idea) is behaviour as if the slope was actually a slope. However with a cubic terrain with flat terraces, real water actually would spread outwards equally in every direction on each flat surface, as in MT.

The reason MT causes ridiculous floods is because it doesn't reduce the volume of water as it spreads, every time water falls off an edge it behaves like the initial water source, so water volume multiplies as it spreads down a hill.

Real water on cubic terraces would spread roughly as in MT but with rapidly falling volume as it spreads.

**Comment by MarkuBu** [comment]
You can't reduce the volume because you have an infinite source and you don't know how far the water can flow. What needs to be limited is the spread. 

It is ok if a source spreads in all directions on a plain surface, as long as there is no drop somewhere.

Each liquid node can spread to each direction horizontally, but if a neighbour is lower than the current node it only spreads to this direction.

You need to check first if the liquid can drop on one or more sides. If not, spread as usual, if yes, only spread to these directions. That's the basic idea. But it could be difficult to implement this to the current code. I don't have the experience in C++ and MT to do this.

This would limit the volume automatically

**Comment by paramat** [comment]
Yes, i support your suggestion (it's how MC does it BTW) as it would reduce the danger from liquid griefs as well as reduce water flow processing.

**Comment by MarkuBu** [comment]
No, MC has a different system. It searches for drops in the whole possibe spreading area and generates only a "stream" to the nearest drop

https://minecraft.gamepedia.com/Liquid#Spread

My suggestion would still spread the water to any direction on plain areas

**Comment by paramat** [comment]
> It searches for drops in the whole possibe spreading area

It seems to search in a 5 node radius which is less than the possible spread area.

> My suggestion would still spread the water to any direction on plain areas

That's what MC does if no drop is found within 5 nodes.

However, we probably could not scan as far as 5 nodes for performance reasons.

**Comment by MarkuBu** [comment]
> However, we probably could not scan as far as 5 nodes for performance reasons.

That's not my suggestion. I just suggest to search only the direct neighbors. The result is somewhere between the current MT behavior and the behavior in MC

**Comment by paramat** [comment]
> Instead of spread to any direction it would make sense that the nodes around are checked if the liquid can flow down. If that is possible on one or more sides the liquid spreads only downwards, if not it spreads like now to all directions.

I see. And only direct neighbours. This would still cause large spreading floods on slopes though. I would prefer the flow is a single flow to the closest drop within 2-4 nodes radius, otherwise normal MT spread.

**Comment by MarkuBu** [comment]
It could still cause larger floods, that's right. But **much** less than now. 

If you want single flow to closest drop than you need to make a single flow to each drop in the same distance, not just the first or random one. I would prefer this too, but makes it more complicated. I think, my suggestion is a good compromise between the desired result and the necessary effort.

I'm not sure, but as far as I understand the code it should be possible to modify the existing code for someone who understands how it works. Single streams to the closest drop would need more code changes I think.

**Comment by paramat** [comment]
> If you want single flow to closest drop than you need to make a single flow to each drop in the same distance, not just the first or random one

I would actually choose the first or random one, again to reduce multiplication of water.

**Comment by o01eg** [comment]
What about use liquid system from Freeminer? https://github.com/minetest/minetest/issues/6144

**Comment by paramat** [comment]
Hehe, NO :) It's actually a nightmare implementation and was one of the reasons the Freeminer fork started, because we refused that liquid system in MT.

**Comment by grapereader** [comment]
> But it is still a valid concern because it is undesired gameplay for some subgames. Maybe one possibility to improve this is to add an internal “liquid parameter” (maybe in param2?) to indicate the total flowing distance from the nearest liquid source feeding into it. Flowing liquids could then simply stop flowing if the total flowing distance is exceeded, even if they would otherwise continue to flow under the normal rules.

Yes! Absolutely the way to prevent these sorts of griefs.

**Comment by 0siribix** [comment]
I am interested in working on this. The information in this discussion is... kind of hard to follow so here is what I'm interested in doing:
Add a server variable that controls whether a water source will spread with air directly beneath it
Add a server variable that controls whether water sources spawn new water sources (Is this already a thing and I missed it?)
Add a finite water source that flows a variable number of blocks no matter the direction. This is what will be given when picked up by a bucket
I don't see the problem with water spreading when it drops on other water but it should be feasible to add a variable that controls how far water spreads when it flows down on another water source

**Comment by Wuzzy2** [comment]
I suggest to close this issue because there is no actual suggestion here, just vague complaints about a lack in feature-richness. This issue is neither a bug report nor a proper feature request.

While some suggestions were made in this issue as comments, they are not part of OP's post and everything is just spread out everywhere and it is chatic. There is no definition about when this issue can be considered "done" because there will always be something that you can "improve" about liquids.

Instead, I suggest to open a fresh new issue with clear, well-defined and understandable suggestions about what liquids should be able to do and how.

**Comment by 0siribix** [comment]
I am interested in working on this. The information in this discussion is... kind of hard to follow so here is what I'm interested in doing:
Add a server variable that controls whether a water source will spread with air directly beneath it
Add a server variable that controls whether water sources spawn new water sources (Is this already a thing and I missed it?)
Add a finite water source that flows a variable number of blocks no matter the direction. This is what will be given when picked up by a bucket
I don't see the problem with water spreading when it drops on other water but it should be feasible to add a variable that controls how far water spreads when it flows down on another water source

**Comment by 0siribix** [comment]
> Instead, I suggest to open a fresh new issue with clear, well-defined and understandable suggestions about what liquids should be able to do and how.

That makes sense

**Comment by Pantyhose-X** [comment]
I found an interesting video about water in Minecraft. Mine test may need to refactor physics and particles
https://www.youtube.com/watch?v=kf7DAWcjmvQ

**Comment by CallMeDax** [comment]
Minetest is many things, but none of these things are Blender.
